Move to Python 3.7+ code

pmhahn opened the pull request:
This addresses #227 to remove the old Python 2.7 compatibility code and moved the code base up to Python 3.7+
This is only lightly tested; `vnclog` still does not work.
Please have a look and review carefully; any (early) feedback from you is very much appreciated. Hopefully I will find time during the Christmas holidays to work on that.
If this is too much I can try to split this into separate PRs, but that is not trivial as some of those patches depend on each other: the code might not `git bisect` at all because some intermediate steps might be "broken".

sibson commented on tests/functional/test_screen.py:
```diff
@@ -2,16 +2,21 @@
 EXAMPLE_PNG = os.path.join(DATADIR, 'example.png')
 EXAMPLE_NOCURSOR_PNG = os.path.join(DATADIR, 'example_nocursor.png')
 
+SERVER = "exmaple"
```
typo?

pmhahn replied:
Yes, will fix.
While at it maybe you can clarify how that  _functional_ unit test worked: It looks like you have an executable named `example` in your test  environment as this is `pexcept.spawn()`ed. Maybe it would be more useful to make that executable configurable, e.g. get it from some environment variable via `os.environ.get("VNCDOTOOL_TEST_SERVER", "example")`

sibson replied:
+1, will add some notes on testing and have a go at making them work in CI too.

sibson commented on tests/unit/helpers.py:
```diff
@@ -2,7 +2,7 @@
 from functools import wraps
 
-class _isolate(object):
+class _isolate:
```
fyi, I'm working on removing all this isolate stuff

sibson replied:
I think I'll rebase my work over this, as it also was doing the mock changes.

sibson commented on vncdotool/command.py:
```diff
@@ -16,15 +15,20 @@
 from .client import VNCDoToolFactory, VNCDoToolClient, factory_connect
 from .loggingproxy import VNCLoggingServerFactory
 
+V = TypeVar("V", bound="VNCDoCLIClient")
```
Is using V an common idiom?  What is this actually doing?

pmhahn replied:
`V`, `K`, `VT`, `VK` are often used for _unbound_ `TypeVar`s, but as now it is _bound_ `V` is no longer a good name. Thank you for spotting this. I changed it to `vncdotool.client.TClient`.

[TypeVar](https://mypy.readthedocs.io/en/stable/generics.html#generics)s are used to bind some type to a name: Take this simple function for an example: `def same(var: V) -> V: return var`
You can pass any type (int, str, object, ….) into that function and will get the same type back.
The additional `bound="VNCDoCLIClient"` adds one more restriction: the passed in value must be an instance of `VNCDoCLIClient` (or any subclass of it) and you will (still) get the same type back. This is important if you actually pass in a sub-class of `VNCDoCLIClient` as then you're guaranteed to get back that sub-class, and not only the super-class as you would get with `def f(var: VNCDoCLIClient) -> VNCDoCLIClient: ...`: There you could still pass in a sub-class of `VNCDoCLIClient`, but you're only guaranteed to get back a `VNCDoCLIClient`, which might then not have the added methods of your sub-class.

sibson reviewed (approved):
Generally LGTM and few clarifying questions.    

How have you tested this?  

I'm wondering if this should be a 2.0 type change.  I'm also thinking that when this goes in I'll close any open and stale PRs because most likely they'll need updating due to conflicts.  What do you think?

pmhahn commented:
> How have you tested this?

You got me: I'm still on it which is why I called this PR a `WIP`. I only did some simple `vncdotool` testing so far.
My company [Univention GmbH](https://www.univention.de/) is using it for our automated testing, which is why I'm working on this at all. Our tests where failing after switching to Python 3 so I had to go down very deep. I had to spend many hours on understanding the low-level details and adding type-hints helped me to understand all the relations.

> I'm wondering if this should be a 2.0 type change. I'm also thinking that when this goes in I'll close any open and stale PRs because most likely they'll need updating due to conflicts. What do you think?

I tried to not break the ABI, but all those `bytes` vs. `unicode` differences might have changed details, especially my commit 1ecaafc9d107 named `fix(rfb): Missing encoding/decoding`: Previously it was _unspecified_ if you still work with low-level `bytes` or already with high-level `unicode`.
The change to use `bytearray()` to buffer the packages should be an internal change and not affect any using application.
The rest is mostly refactoring and style changes, which would not require a major version bump. But I'm okay with this.

I only have looked shortly at the pending PRs: Most of them will no longer work because adding PEP-484 type annotations changed most context. If you take them before this PR I'll rework my branch to incorporate them, but the other way around would also be okay.

pmhahn commented:
Last run looks better, but the [build for 3.11 is failing](https://github.com/univention/vncdotool/actions/runs/3757803532/jobs/6385410415) because my branch (based on `main`) still uses `isolate` while your branch `master` has it already removed.
I cherry-picked your patches 9ca371f8da99399e553eb10f230b509eb2601d38 3c17a6583a403e3730db6187cbfcee1db3fdea04 30d4db5db9c41f6634b611939fb3d9fef1474a4c 84de3c922b03b21fd95ea8c44ae1503f19e34651 and squashed them into one and fixed `tests/unit/test_command.py`.

sibson commented:
Let me know when you think this is mostly good enough to merge.  I think we should merge sooner rather than later.  It will be easier for us to make additional cleanups against main than holding a PR open.

pmhahn commented:
From my PoV it's ready; I undid my change to `pytest` as you already merged your change to plain `unittest` into `main`.
My testing so far found no issues, but  your eyes and testing as the expert are much appreciated.

pmhahn commented:
> > I'm wondering if this should be a 2.0 type change. I'm also thinking that when this goes in I'll close any open and stale PRs because most likely they'll need updating due to conflicts. What do you think?
> 
> I tried to not break the ABI

Removing Python 2.x support is a breaking change, so using Semantic versioning the version should be `2.0.0`.

sibson commented:
huge thanks @pmhahn for putting all this together.

sibson commented:
@pmhahn fyi, https://github.com/sibson/vncdotool/commit/54bcdb23a9e0c560774009680774041af07038f5, sets up the environment to run the functional tests.  They are currently failing due to a bytes/string issue.

pmhahn commented:
> @pmhahn fyi, [54bcdb2](https://github.com/sibson/vncdotool/commit/54bcdb23a9e0c560774009680774041af07038f5), sets up the environment to run the functional tests. They are currently failing due to a bytes/string issue.

Thanks for that; see #231 which fixes those bytes/string issues.
